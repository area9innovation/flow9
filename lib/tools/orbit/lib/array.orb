// Array utility functions library for Orbit

// map: Apply a function to each element in an array
fn map(arr, f) = (
    if length(arr) = 0 then []
    else [f(arr[0])] + map(subrange(arr, 1, length(arr) - 1), f)
);

// filter: Keep only elements that match a predicate
fn filter(arr, pred) = (
    if length(arr) = 0 then []
    else if pred(arr[0]) then 
        [arr[0]] + filter(subrange(arr, 1, length(arr) - 1), pred)
    else 
        filter(subrange(arr, 1, length(arr) - 1), pred)
);

// fold: Accumulate a result based on each array element and an accumulator
fn fold(arr, init, f) = (
    if length(arr) = 0 then init
    else fold(subrange(arr, 1, length(arr) - 1), f(init, arr[0]), f)
);

// mapi: Map with index
fn mapi(arr, f) = (
    fn mapiHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else [f(arr[0], idx)] + mapiHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
    );
    mapiHelper(arr, f, 0)
);

// filteri: Filter with index
fn filteri(arr, pred) = (
    fn filteriHelper(arr, pred, idx) = (
        if length(arr) = 0 then []
        else if pred(arr[0], idx) then 
            [arr[0]] + filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
        else 
            filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
    );
    filteriHelper(arr, pred, 0)
);

// foldi: Fold with index
fn foldi(arr, init, f) = (
    fn foldiHelper(arr, acc, f, idx) = (
        if length(arr) = 0 then acc
        else foldiHelper(subrange(arr, 1, length(arr) - 1), f(acc, arr[0], idx), f, idx + 1)
    );
    foldiHelper(arr, init, f, 0)
);

// tail: Return all elements except the first one
fn tail(arr) = (
    if length(arr) <= 1 then []
    else subrange(arr, 1, length(arr) - 1)
);

// tailFrom: Return all elements from a given index to the end
fn tailFrom(arr, idx) = (
    if idx >= length(arr) then []
    else subrange(arr, idx, length(arr) - idx)
);

// take: Return the first n elements of an array
fn take(arr, n) = (
    if n <= 0 || length(arr) = 0 then []
    else [arr[0]] + take(subrange(arr, 1, length(arr) - 1), n - 1)
);

// iter: Call a function for each element (for side effects)
fn iter(arr, f) = (
    if length(arr) = 0 then 0 // Return 0 instead of ()
    else (
        f(arr[0]);
        iter(subrange(arr, 1, length(arr) - 1), f)
    )
);

// iteri: Iterate with index
fn iteri(arr, f) = (
    fn iteriHelper(arr, f, idx) = (
        if length(arr) = 0 then 0 // Return 0 instead of ()
        else (
            f(arr[0], idx);
            iteriHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
        )
    );
    iteriHelper(arr, f, 0)
);

// filtermap: Combine filter and map operations
fn filtermap(arr, f) = (
    if length(arr) = 0 then []
    else (
        let result = f(arr[0]);
        result is (
            Some(val) => [val] + filtermap(subrange(arr, 1, length(arr) - 1), f);
            None() => filtermap(subrange(arr, 1, length(arr) - 1), f);
            _ => filtermap(subrange(arr, 1, length(arr) - 1), f) // For safety
        )
    )
);

// filtermapi: Filtermap with index
fn filtermapi(arr, f) = (
    fn filtermapHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else (
            let result = f(arr[0], idx);
            result is (
                Some(val) => [val] + filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                None() => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                _ => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1) // For safety
            )
        )
    );
    filtermapHelper(arr, f, 0)
);

// removeIndex: Remove an element at a specific index
fn removeIndex(arr, idx) = (
    if idx < 0 || idx >= length(arr) then arr
    else take(arr, idx) + tailFrom(arr, idx + 1)
);

// removeRange: Remove a range of elements
fn removeRange(arr, start, count) = (
    if start < 0 || start >= length(arr) || count <= 0 then arr
    else take(arr, start) + tailFrom(arr, start + count)
);

// removeFirst: Remove the first occurrence of a specific element
fn removeFirst(arr, element) = (
    fn removeFirstHelper(before, after) = (
        if length(after) = 0 then before
        else if after[0] = element then before + subrange(after, 1, length(after) - 1)
        else removeFirstHelper(before + [after[0]], subrange(after, 1, length(after) - 1))
    );
    removeFirstHelper([], arr)
);

// removeAll: Remove all occurrences of a specific element
fn removeAll(arr, element) = (
    filter(arr, \x.x != element != 0)
);

// contains: Check if an element exists in the array
fn contains(arr, element) = (
    if length(arr) = 0 then false
    else if arr[0] = element then true
    else contains(subrange(arr, 1, length(arr) - 1), element)
);

// exists: Check if any element satisfies a predicate
fn exists(arr, pred) = (
    if length(arr) = 0 then false
    else if pred(arr[0]) then true
    else exists(subrange(arr, 1, length(arr) - 1), pred)
);

// forall: Check if all elements satisfy a predicate
fn forall(arr, pred) = (
    if length(arr) = 0 then true
    else if !pred(arr[0]) then false
    else forall(subrange(arr, 1, length(arr) - 1), pred)
);

// uniq: Remove duplicates from an array
fn uniq(arr) = (
    if length(arr) <= 1 then arr
    else (
        fn uniq_helper(remaining, result) = (
            if length(remaining) = 0 then result
            else (
                let current = remaining[0];
                let rest = tail(remaining);
                
                if contains(result, current) then
                    // Skip duplicate
                    uniq_helper(rest, result)
                else
                    // Keep this element
                    uniq_helper(rest, result + [current])
            )
        );
        uniq_helper(arr, [])
    )
);