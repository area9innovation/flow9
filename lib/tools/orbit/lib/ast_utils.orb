// AST utility functions for Orbit

// Create a properly nested AST node by folding over an array of arguments
// This function handles arrays of any length, whereas makeAst expects exactly 2 arguments
fn makeFoldedAst(op, args) = (
    let len = length(args);
    
    if len = 0 then (
        // Handle empty array case - return an appropriate default value
        // This depends on the operator, but we'll return a neutral element
        if op = "&&" then true
        else if op = "||" then false
        else args  // Return empty array as fallback
    ) else if len = 1 then (
        // For single element, just return it without creating an operation
        args[0]
    ) else if len = 2 then (
        // For exactly two elements, we can use makeAst directly
        makeAst(op, args)
    ) else (
        // For 3+ elements, fold to create a proper binary tree
        // Start with the first element and fold over the rest
        let initialAst = args[0];
        let remainingArgs = subrange(args, 1, len - 1);
        
        // Fold over the remaining arguments, combining each with the accumulator
        fold(remainingArgs, initialAst, \acc, element -> (
            makeAst(op, [acc, element])
        ))
    )
);