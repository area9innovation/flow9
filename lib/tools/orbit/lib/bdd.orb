import lib/rewrite;
import lib/array;
import lib/logic;

/**
 * Binary Decision Diagrams (BDDs)
 * 
 * This library implements Binary Decision Diagrams (BDDs) for Boolean formulas,
 * based on Shannon's expansion theorem. BDDs provide a canonical representation
 * for Boolean functions, which makes equivalence checking efficient.
 * 
 * BDDs are represented using the if-then-else (Ite) constructor:
 * Ite(var, then_branch, else_branch)
 * 
 * Where:
 * - var is a decision variable
 * - then_branch is the BDD when var is true
 * - else_branch is the BDD when var is false
 * - Terminal nodes are true and false
 */

// ----- VARIABLE ORDERING -----

// Get a numeric value for a variable to determine ordering
fn var_order(var) = (
    if is_not(var) then var_order(get_not_arg(var))
    else var
);

// Check if an expression is a negation
fn is_not(expr) = (
    getConstructor(expr) == "!"
);

// Get the argument of a negation
fn get_not_arg(expr) = (
    expr is (!arg => arg; _ => expr)
);

// ----- BDD CONSTRUCTION -----

// Convert a Boolean formula to a BDD
fn to_bdd(expr) = (
    expr is (
        // Terminal cases
        true => true;
        false => false;
        
        // Negation
		¬e => negate_bdd(to_bdd(e));
        
        // Conjunction (AND)
        a ∧ b => apply_and(to_bdd(a), to_bdd(b));
        
        // Disjunction (OR)
		a ∨ b => apply_or(to_bdd(a), to_bdd(b));
        
        // Already a BDD - ensure consistency
        Ite(v, t, e) => make_ite(v, to_bdd(t), to_bdd(e));
        
        // Any other expression type
        // Variable becomes a simple BDD
        v => if is_var(v) then make_ite(v, true, false) else expr;
    )
);

// Check if an expression is a variable
fn is_var(expr) = (
    let c = getConstructor(expr);
    c != "Ite" && c != "!" && c != "&&" && c != "||" && 
    expr != true && expr != false
);

// Create an Ite node with reduction
fn make_ite(var, then_case, else_case) = (
    // Rule 1: If both branches are equal, eliminate the test
    if then_case == else_case then then_case
    else Ite(var, then_case, else_case)
);

// Negate a BDD
fn negate_bdd(bdd) = (
    bdd is (
        true => false;
        false => true;
        Ite(v, t, e) => make_ite(v, negate_bdd(t), negate_bdd(e));
        _ => !bdd  // For variables or unprocessed expressions
    )
);

// Apply AND operation to two BDDs
fn apply_and(bdd1, bdd2) = (
    // Terminal cases
    if bdd1 == false || bdd2 == false then false
    else if bdd1 == true then bdd2
    else if bdd2 == true then bdd1
    else bdd1 is (
        Ite(v1, t1, e1) => (
            bdd2 is (
                Ite(v2, t2, e2) => (
                    // Choose variable to expand based on ordering
                    let v_order1 = var_order(v1);
                    let v_order2 = var_order(v2);
                    if v_order1 == v_order2 then (
                        // Same variable - combine results
                        make_ite(v1, apply_and(t1, t2), apply_and(e1, e2))
                    ) else if v_order1 < v_order2 then (
                        // v1 comes first - expand on v1
                        make_ite(v1, apply_and(t1, bdd2), apply_and(e1, bdd2))
                    ) else (
                        // v2 comes first - expand on v2
                        make_ite(v2, apply_and(bdd1, t2), apply_and(bdd1, e2))
                    )
                );
                // bdd2 is not an Ite node
                _ => make_ite(v1, apply_and(t1, bdd2), apply_and(e1, bdd2))
            )
        );
        // bdd1 is not an Ite node
        _ => bdd2 is (
            Ite(v2, t2, e2) => make_ite(v2, apply_and(bdd1, t2), apply_and(bdd1, e2));
            _ => bdd1 && bdd2  // Neither is an Ite node
        )
    )
);

// Apply OR operation to two BDDs
fn apply_or(bdd1, bdd2) = (
    // Terminal cases
    if bdd1 == true || bdd2 == true then true
    else if bdd1 == false then bdd2
    else if bdd2 == false then bdd1
    else bdd1 is (
        Ite(v1, t1, e1) => (
            bdd2 is (
                Ite(v2, t2, e2) => (
                    // Choose variable to expand based on ordering
                    let v_order1 = var_order(v1);
                    let v_order2 = var_order(v2);
                    
                    if v_order1 == v_order2 then (
                        // Same variable - combine results
                        make_ite(v1, apply_or(t1, t2), apply_or(e1, e2))
                    ) else if v_order1 < v_order2 then (
                        // v1 comes first - expand on v1
                        make_ite(v1, apply_or(t1, bdd2), apply_or(e1, bdd2))
                    ) else (
                        // v2 comes first - expand on v2
                        make_ite(v2, apply_or(bdd1, t2), apply_or(bdd1, e2))
                    )
                );
                // bdd2 is not an Ite node
                _ => make_ite(v1, apply_or(t1, bdd2), apply_or(e1, bdd2))
            )
        );
        // bdd1 is not an Ite node
        _ => bdd2 is (
            Ite(v2, t2, e2) => make_ite(v2, apply_or(bdd1, t2), apply_or(bdd1, e2));
            _ => bdd1 || bdd2  // Neither is an Ite node
        )
    )
);

// ----- BDD TO LOGIC CONVERSION -----

fn makeOr(a, b) = (
	scatter(`||`([a,b]))
);

fn makeAnd(a, b) = (
	scatter(`&&`([a,b]))
);

fn makeNot(a) = (
	makeAst("!", ([a]))
);

// Convert a BDD back to a standard Boolean formula
fn from_bdd(bdd) = (
    bdd is (
        // Terminal cases
        true => true;
        false => false;
        
        // BDD node
        Ite(v, t, e) => (
            let then_expr = from_bdd(t);
            let else_expr = from_bdd(e);
            
            // Simplify common cases
            if then_expr == true && else_expr == false then v
            else if then_expr == false && else_expr == true then makeNot(v)
            else if then_expr == true then makeOr(v, else_expr)
            else if else_expr == false then makeAnd(v, then_expr)
			else if else_expr == true then makeOr(makeNot(v), then_expr)
            else if then_expr == false then makeAnd(makeNot(v), else_expr)
            else (
				makeOr(
					makeAnd(v, then_expr),
					makeAnd(makeNot(v), else_expr)
				)
			)
        );
        
        // Variable or other expression
        _ => bdd
    )
);

// ----- VARIABLE EXTRACTION -----

// Extract all variables from a logic expression in order
fn extract_variables(expr) = (
    expr is (
        // Terminal cases
        true => [];
        false => [];
        
        // Variable
        v => if is_var(v) then [v] else [];
        
        // Negation
        !e => extract_variables(e);
        
        // AND/OR - combine variables from both sides
        a && b => uniq(sort(extract_variables(a) + extract_variables(b)));
        a || b => uniq(sort(extract_variables(a) + extract_variables(b)));
        
        // BDD node
        Ite(v, t, e) => uniq(sort([v] + extract_variables(t) + extract_variables(e)));
        
        // Fallback
        _ => []
    )
);

// ----- UTILITY FUNCTIONS -----

// Evaluate a BDD with a variable assignment
fn evaluate_bdd(bdd, assignment) = (
    bdd is (
        // Terminal cases
        true => true;
        false => false;
        
        // BDD node - follow appropriate branch based on variable value
        Ite(v, t, e) => (
            let value = assignment(v);
            if value then evaluate_bdd(t, assignment) else evaluate_bdd(e, assignment)
        );
        
        // Variable or unprocessed expression
        _ => assignment(bdd)
    )
);

// ----- EQUIVALENCE CHECKING -----

// Check if two Boolean formulas are equivalent
fn are_equivalent(expr1, expr2) = (
    let bdd1 = to_bdd(expr1);
    let bdd2 = to_bdd(expr2);
    bdd1 == bdd2
);

// ----- QUANTIFICATION OPERATIONS -----

// Variable restriction: substitute variable with a value
fn bdd_restrict(node, var, value) = (
    // Terminal cases are unchanged by restriction
    if node == true || node == false then node
    else node is (
        Ite(v, t, e) => (
            if v == var then
                // Direct substitution when we find the variable
                if value then t else e
            else (
                // For other variables, recurse on both branches
                make_ite(v, bdd_restrict(t, var, value), bdd_restrict(e, var, value))
            )
        );
        // For variables or other expressions
        _ => if node == var then value else node
    )
);

// Existential quantification: ∃v.f
fn bdd_exists(var, node) = (
    // Compute f[v→0] ∨ f[v→1]
    apply_or(
        bdd_restrict(node, var, false),
        bdd_restrict(node, var, true)
    )
);

// Universal quantification: ∀v.f
fn bdd_forall(var, node) = (
    // Compute f[v→0] ∧ f[v→1]
    apply_and(
        bdd_restrict(node, var, false),
        bdd_restrict(node, var, true)
    )
);

// Apply existential quantification for multiple variables
fn bdd_exists_multiple(vars, node) = (
    if length(vars) == 0 then node
    else bdd_exists_multiple(
        tail(vars), 
        bdd_exists(head(vars), node)
    )
);

// Apply universal quantification for multiple variables
fn bdd_forall_multiple(vars, node) = (
    if length(vars) == 0 then node
    else bdd_forall_multiple(
        tail(vars), 
        bdd_forall(head(vars), node)
    )
);

// ----- MODEL COUNTING -----

// Calculate the number of satisfying assignments for a BDD
// Simple implementation without using cache
fn bdd_count_models(node, var_count) = (
    node is (
        // No satisfying assignments for false
        false => 0;
        
        // All remaining variable assignments satisfy true
        true => (
            // 2^remaining_vars
            let result = 1;
            fn calculate_power(base, exponent, acc) = (
                if exponent <= 0 then acc
                else calculate_power(base, exponent - 1, acc * base)
            );
            calculate_power(2, var_count, 1)
        );
        
        // For BDD nodes, compute recursively
        Ite(v, t, e) => (
            // Count satisfying assignments in both branches
            let low_count = bdd_count_models(e, var_count - 1);
            let high_count = bdd_count_models(t, var_count - 1);
            
            // Sum the counts
            low_count + high_count
        );
        
        // For variables or other expressions
        _ => if node then (
            // 2^var_count
            let result = 1;
            fn calculate_power(base, exponent, acc) = (
                if exponent <= 0 then acc
                else calculate_power(base, exponent - 1, acc * base)
            );
            calculate_power(2, var_count, 1)
        ) else 0
    )
);

// Find one satisfying assignment for a BDD
fn bdd_find_satisfying_assignment(node) = (
    if node == false then
        None()  // No satisfying assignment
    else if node == true then
        Some([])  // Empty assignment works
    else node is (
        Ite(v, t, e) => (
            // Try high branch first (v=true)
            let high_result = bdd_find_satisfying_assignment(t);
            high_result is (
                Some(assignment) => Some([Pair(v, true)] + assignment);
                None() => (
                    // Try low branch (v=false)
                    let low_result = bdd_find_satisfying_assignment(e);
                    low_result is (
                        Some(assignment) => Some([Pair(v, false)] + assignment);
                        None() => None()
                    )
                )
            )
        );
        // For variables, assign true
        _ => Some([Pair(node, true)])
    )
);