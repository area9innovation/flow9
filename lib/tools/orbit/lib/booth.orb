import lib/array;

// Linear time Booth's algorithm to find canonical form in cyclic groups
fn booth_canonical(arr) = (
	if length(arr) <= 1 then arr
	else (
		let n = length(arr);

		// Create an auxiliary array which is the original array concatenated with itself
		let double_arr = arr + arr;

		// Initialize failure function array with -1
		// (We use -1 instead of conventional 0 to simplify logic)
		let f = \i. if i < 0 || i >= n then -1 else f(i);

		// Compute the min rotation index in linear time
		fn compute_min_index() = (
			// Initialize starting values
			fn find_min(i, j, k) = (
				// We're done when we've checked all rotations
				if i + k >= 2 * n then j
				else if k >= n then find_min(i + 1, j, 0)
				else (
					// Compute the current indices we're comparing
					let i_idx = (i + k) % (2 * n);
					let j_idx = (j + k) % (2 * n);

					if double_arr[i_idx] < double_arr[j_idx] then
						// i rotation is smaller, j becomes i
						find_min(i, i, k + 1)
					else if double_arr[i_idx] > double_arr[j_idx] then
						// j rotation is smaller, i moves forward
						find_min(i + k + 1, j, 0)
					else
						// Equal so far, continue comparison
						find_min(i, j, k + 1)
				)
			);

			// Start with first two positions
			find_min(0, 0, 0)
		);

		let min_start = compute_min_index();

		// Build the canonical form starting from min_start
		fn build_result(idx, result) = (
			if idx >= n then result
			else build_result(idx + 1, result + [arr[(min_start + idx) % n]])
		);

		build_result(0, [])
	)
);

