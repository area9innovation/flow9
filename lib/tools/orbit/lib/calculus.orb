// File: /home/alstrup/area9/flow9/lib/tools/orbit/lib/calculus.orb
//
// Library for symbolic Calculus operations in Orbit.
// Includes rules for Differentiation, Integration, and Limits.

import lib/rewrite;
import lib/reflect;
import lib/complex; // For 'i' if needed, and Complex constants

// Define mathematical constants if not universally available
let pi = 3.141592653589793;
let e = 2.718281828459045;

// Helper for conditions
fn is_numeric(x) = (isInt(x) || isDouble(x) || isConstructor(x) && getConstructor(x) == "Complex");
fn is_constant(expr, var) = (!containsVar(expr, var)); // Assumes containsVar helper exists

// --- Differentiation Rules ---
// Using 'diff(expr, var)' operator

let differentiation_rules = [
    // Basic Rules
    prepareRule("diff_constant", quote(diff(c, x)), quote(0), is_constant(c, x)),
    prepareRule("diff_variable_self", quote(diff(x, x)), quote(1), true),
    prepareRule("diff_variable_other", quote(diff(y, x)), quote(0), eval(x != y) && is_var(y)), // If y is a variable different from x
    prepareRule("diff_sum", quote(diff(f + g, x)), quote(diff(f, x) + diff(g, x)), true),
    prepareRule("diff_sub", quote(diff(f - g, x)), quote(diff(f, x) - diff(g, x)), true),
    prepareRule("diff_scalar_mul", quote(diff(c * f, x)), quote(c * diff(f, x)), is_constant(c, x)),
    prepareRule("diff_product", quote(diff(f * g, x)), quote(f * diff(g, x) + g * diff(f, x)), true),
    prepareRule("diff_quotient", quote(diff(f / g, x)), quote((g * diff(f, x) - f * diff(g, x)) / pow(g, 2.0)), true), // Use pow for g^2
    prepareRule("diff_chain", quote(diff(f(g), x)), quote(diff(f, g) * diff(g, x)), true), // Conceptual: diff(f,g) needs care

    // Power Rule
    prepareRule("diff_power", quote(diff(f^n, x)), quote(n * pow(f, n - 1.0) * diff(f, x)), is_numeric(n)), // Use pow

    // Exponential Functions
    prepareRule("diff_exp_x", quote(diff(exp(x), x)), quote(exp(x)), true),
    prepareRule("diff_exp_f", quote(diff(exp(f), x)), quote(exp(f) * diff(f, x)), true), // Chain rule applied
    prepareRule("diff_a_pow_x", quote(diff(pow(a, x), x)), quote(pow(a, x) * log(a)), is_constant(a, x)), // Use pow
    prepareRule("diff_a_pow_f", quote(diff(pow(a, f), x)), quote(pow(a, f) * log(a) * diff(f, x)), is_constant(a, x)), // Use pow

    // Logarithmic Functions
    prepareRule("diff_log_x", quote(diff(log(x), x)), quote(1.0 / x), true), // Natural log
    prepareRule("diff_log_f", quote(diff(log(f), x)), quote((1.0 / f) * diff(f, x)), true), // Chain rule applied
    prepareRule("diff_log_base_a_x", quote(diff(log_base(a, x), x)), quote(1.0 / (x * log(a))), is_constant(a, x)),
    prepareRule("diff_log_base_a_f", quote(diff(log_base(a, f), x)), quote((1.0 / (f * log(a))) * diff(f, x)), is_constant(a, x)),

    // Trigonometric Functions
    prepareRule("diff_sin_x", quote(diff(sin(x), x)), quote(cos(x)), true),
    prepareRule("diff_sin_f", quote(diff(sin(f), x)), quote(cos(f) * diff(f, x)), true),
    prepareRule("diff_cos_x", quote(diff(cos(x), x)), quote(0.0 - sin(x)), true), // Use subtraction
    prepareRule("diff_cos_f", quote(diff(cos(f), x)), quote((0.0 - sin(f)) * diff(f, x)), true),
    prepareRule("diff_tan_x", quote(diff(tan(x), x)), quote(pow(sec(x), 2.0)), true), // Use pow
    prepareRule("diff_tan_f", quote(diff(tan(f), x)), quote(pow(sec(f), 2.0) * diff(f, x)), true),
    prepareRule("diff_sec_x", quote(diff(sec(x), x)), quote(sec(x) * tan(x)), true),
    prepareRule("diff_sec_f", quote(diff(sec(f), x)), quote(sec(f) * tan(f) * diff(f, x)), true),

    // Inverse Trigonometric Functions
    prepareRule("diff_asin_x", quote(diff(asin(x), x)), quote(1.0 / sqrt(1.0 - pow(x, 2.0))), true),
    prepareRule("diff_acos_x", quote(diff(acos(x), x)), quote(-1.0 / sqrt(1.0 - pow(x, 2.0))), true),
    prepareRule("diff_atan_x", quote(diff(atan(x), x)), quote(1.0 / (1.0 + pow(x, 2.0))), true)
];

// --- Integration Rules (Indefinite) ---
// Using 'integrate(expr, var)' operator. Handling the constant 'C' is tricky in rewriting.
// These rules represent antiderivatives; a full symbolic integrator is much more complex.

let integration_rules = [
    // Basic Rules
    prepareRule("integrate_constant", quote(integrate(c, x)), quote(c * x), is_constant(c, x)),
    prepareRule("integrate_sum", quote(integrate(f + g, x)), quote(integrate(f, x) + integrate(g, x)), true),
    prepareRule("integrate_sub", quote(integrate(f - g, x)), quote(integrate(f, x) - integrate(g, x)), true),
    prepareRule("integrate_scalar_mul", quote(integrate(c * f, x)), quote(c * integrate(f, x)), is_constant(c, x)),

    // Power Rule (n != -1)
    prepareRule("integrate_power", quote(integrate(pow(x, n), x)), quote(pow(x, n + 1.0) / (n + 1.0)), is_numeric(n) && eval(n != -1.0)),

    // Special case n = -1 (1/x)
    prepareRule("integrate_one_over_x", quote(integrate(pow(x, -1.0), x)), quote(log(abs(x))), true),
    prepareRule("integrate_one_over_x_alt", quote(integrate(1.0 / x, x)), quote(log(abs(x))), true),

    // Exponential Functions
    prepareRule("integrate_exp_x", quote(integrate(exp(x), x)), quote(exp(x)), true),
    prepareRule("integrate_a_pow_x", quote(integrate(pow(a, x), x)), quote(pow(a, x) / log(a)), is_constant(a, x) && eval(a > 0.0 && a != 1.0)),

    // Trigonometric Functions
    prepareRule("integrate_sin_x", quote(integrate(sin(x), x)), quote(0.0 - cos(x)), true),
    prepareRule("integrate_cos_x", quote(integrate(cos(x), x)), quote(sin(x)), true),
    prepareRule("integrate_sec_squared_x", quote(integrate(pow(sec(x), 2.0), x)), quote(tan(x)), true),
    prepareRule("integrate_sec_tan_x", quote(integrate(sec(x) * tan(x), x)), quote(sec(x)), true),

    // Inverse Trig Forms
    prepareRule("integrate_inv_sqrt_one_minus_x_sq", quote(integrate(1.0 / sqrt(1.0 - pow(x, 2.0)), x)), quote(asin(x)), true),
    prepareRule("integrate_inv_one_plus_x_sq", quote(integrate(1.0 / (1.0 + pow(x, 2.0)), x)), quote(atan(x)), true)

    // NOTE: Symbolic integration often requires pattern matching for u-substitution,
    // integration by parts, partial fractions etc., which are complex to express solely
    // as simple rewrite rules without a more sophisticated strategy engine.
];

// --- Limit Rules ---
// Using 'limit(expr, var, tends_to)' operator. 'tends_to' can be a value or 'infinity'.

let limit_rules = [
    // Basic Limits
    prepareRule("limit_constant", quote(limit(c, x, a)), quote(c), is_constant(c, x)),
    prepareRule("limit_variable", quote(limit(x, x, a)), quote(a), true),
    prepareRule("limit_scalar_mul", quote(limit(c * f, x, a)), quote(c * limit(f, x, a)), is_constant(c, x)),
    prepareRule("limit_sum", quote(limit(f + g, x, a)), quote(limit(f, x, a) + limit(g, x, a)), true),
    prepareRule("limit_product", quote(limit(f * g, x, a)), quote(limit(f, x, a) * limit(g, x, a)), true),
    prepareRule("limit_quotient", quote(limit(f / g, x, a)), quote(limit(f, x, a) / limit(g, x, a)), eval(limit(g, x, a) != 0.0)),
    prepareRule("limit_power", quote(limit(pow(f, g), x, a)), quote(pow(limit(f, x, a), limit(g, x, a))), true), // Requires continuity checks in reality
    prepareRule("limit_composition", quote(limit(f(g), x, a)), quote(f(limit(g, x, a))), true), // Requires continuity of f

    // Limits involving Infinity
    prepareRule("limit_one_over_x_inf", quote(limit(1.0 / x, x, infinity)), quote(0.0), true),
    prepareRule("limit_one_over_x_neg_inf", quote(limit(1.0 / x, x, neg_infinity)), quote(0.0), true), // Assuming neg_infinity symbol
    prepareRule("limit_one_over_x_pow_inf", quote(limit(1.0 / pow(x, n), x, infinity)), quote(0.0), is_numeric(n) && eval(n > 0.0)),
    prepareRule("limit_x_pow_n_inf", quote(limit(pow(x, n), x, infinity)), quote(infinity), is_numeric(n) && eval(n > 0.0)),
    prepareRule("limit_x_pow_n_neg_inf", quote(limit(pow(x, n), x, neg_infinity)), quote(infinity), isInt(n) && eval(n > 0 && n % 2 == 0)), // Even power
    prepareRule("limit_x_pow_n_neg_inf_odd", quote(limit(pow(x, n), x, neg_infinity)), quote(neg_infinity), isInt(n) && eval(n > 0 && n % 2 != 0)), // Odd power

    // Limits of specific functions
    prepareRule("limit_exp_inf", quote(limit(exp(x), x, infinity)), quote(infinity), true),
    prepareRule("limit_exp_neg_inf", quote(limit(exp(x), x, neg_infinity)), quote(0.0), true),
    prepareRule("limit_log_inf", quote(limit(log(x), x, infinity)), quote(infinity), true),
    prepareRule("limit_log_zero_plus", quote(limit(log(x), x, zero_plus)), quote(neg_infinity), true), // Limit from right

    // L'HÃ´pital's Rule (Conceptual - requires checking 0/0 or inf/inf form)
    // prepareRule("lhopital",
    //     quote(limit(f / g, x, a)),
    //     quote(limit(diff(f, x) / diff(g, x), x, a)),
    //     eval(limit(f, x, a) == 0.0 && limit(g, x, a) == 0.0) // Indeterminate form 0/0
    // )

    // Squeeze Theorem (Conceptual)
    // if (limit(f, x, a) == L && limit(h, x, a) == L && f <= g && g <= h)
    // then limit(g, x, a) == L
];


// Helper function (conceptual) - needed for is_constant rule
fn containsVar(expr : ast, var : ast) -> bool = (
    expr is (
        v => eval(v == var) && is_var(v); // Direct match
        a + b => containsVar(a, var) || containsVar(b, var);
        a - b => containsVar(a, var) || containsVar(b, var);
        a * b => containsVar(a, var) || containsVar(b, var);
        a / b => containsVar(a, var) || containsVar(b, var);
        a ^ b => containsVar(a, var) || containsVar(b, var); // Check both base and exponent
        f(arg) => containsVar(f, var) || containsVar(arg, var); // Simple function call
        // Add cases for other expression types as needed
        _ => false // Default: does not contain variable
    )
);


"Calculus Rules Library Loaded"
