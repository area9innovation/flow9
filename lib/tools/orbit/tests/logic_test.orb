import lib/logic;

// TODO: Implement sort and uniq in the logic library

// Test CNF (Conjunctive Normal Form) with duplicates and unsorted terms
let cnf_expr = quote((a || c || b || c) && (b || d || b) && (c || a || c));
let canonical_cnf = to_cnf(cnf_expr);

println("Original CNF expression:");
println(prettyOrbit(cnf_expr));
println("Canonical CNF (sorted and unique):");
println(prettyOrbit(canonical_cnf));

// Test DNF (Disjunctive Normal Form) with duplicates and unsorted terms
let dnf_expr = quote((a && c && b && c) || (b && d && b) || (c && a && c));
let canonical_dnf = to_dnf(dnf_expr);

println("\nOriginal DNF expression:");
println(prettyOrbit(dnf_expr));
println("Canonical DNF (sorted and unique):");
println(prettyOrbit(canonical_dnf));
