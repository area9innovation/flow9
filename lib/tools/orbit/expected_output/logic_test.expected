Original CNF expression:
(((((a ∨ c) ∨ b) ∨ c) ∧ ((b ∨ d) ∨ b)) ∧ ((c ∨ a) ∨ c))
Canonical CNF (sorted and unique):
((((b ∨ d) ∨ b) ∧ ((c ∨ a) ∨ c)) ∧ (((a ∨ c) ∨ b) ∨ c))
Original DNF expression:
(((((a ∧ c) ∧ b) ∧ c) ∨ ((b ∧ d) ∧ b)) ∨ ((c ∧ a) ∧ c))
Canonical DNF (sorted and unique):
((((b ∧ d) ∧ b) ∨ ((c ∧ a) ∧ c)) ∨ (((a ∧ c) ∧ b) ∧ c))
Result: "((((b ∧ d) ∧ b) ∨ ((c ∧ a) ∧ c)) ∨ (((a ∧ c) ∧ b) ∧ c))"
